pkg/lokifrontend/frontend/v2/frontend.go

```diff
@@ -144,19 +146,30 @@ func NewFrontend(cfg Config, ring ring.ReadRing, log log.Logger, reg prometheus.
 }
 
 func (f *Frontend) starting(ctx context.Context) error {
-	return errors.Wrap(services.StartAndAwaitRunning(ctx, f.schedulerWorkers), "failed to start frontend scheduler workers")
+	return errors.Wrap(services.StartAndAwaitRunning(context.Background(), f.schedulerWorkers), "failed to start frontend scheduler workers")
 }
 
-func (f *Frontend) stopping(_ error) error {
+func (f *Frontend) stopping(e error) error {
+	t := time.NewTicker(500 * time.Millisecond)
+	defer t.Stop()
+
+	start := time.Now()
+
+	for range t.C {
+		inflight := f.requests.count()
+		elapsed := time.Since(start)
+		if elapsed > f.cfg.GracefulShutdownTimeout || inflight <= 0 {
+			level.Debug(f.log).Log("msg", "no outstanding queries waiting to complete", "inflight", f.requests.count(), "elapsed", elapsed)
```

Comment:
This log message is not correct. It may be printed when elapsed is larger than  `GracefulShutdownTimeout` even though there may be outstanding inflight requests.

Maybe separate it into two _if_ statements with different logs?

Reply:
Right, I'll revert https://github.com/grafana/loki/pull/7978/commits/09da99f99032f5d4ab9cd5240fd76cf1bf556057 again.